lib/llm/src/block_manager/block/registry.rs

```diff
@@ -34,27 +55,82 @@ pub enum BlockRegistationError {
     InvalidState(String),
 }
 
-/// Error returned when an attempt is made to unregister a block that is still active.
-#[derive(Debug, thiserror::Error)]
-#[error("Failed to unregister block: {0}")]
-pub struct UnregisterFailure(SequenceHash);
+/// A block entry is a handle to a block that is registered in the pool.
+/// On drop, we need to notify the pool that the block has been unregistered.
+/// This is different than the registration handle, which is only dropped when the block is no longer in ANY pool.
+#[derive(Debug)]
+pub struct BlockHandle {
+    sequence_hash: SequenceHash,
+    unregister_tx: mpsc::UnboundedSender<SequenceHash>,
+}
+
+impl BlockHandle {
+    pub fn new(
+        sequence_hash: SequenceHash,
+        unregister_tx: mpsc::UnboundedSender<SequenceHash>,
+    ) -> Self {
+        Self {
+            sequence_hash,
+            unregister_tx,
+        }
+    }
+}
+
+impl Drop for BlockHandle {
+    fn drop(&mut self) {
+        let _ = self.unregister_tx.send(self.sequence_hash);
+    }
+}
 
-#[derive()]
 pub struct BlockRegistry {
-    blocks: HashMap<SequenceHash, Weak<RegistrationHandle>>,
+    blocks: Arc<Mutex<HashMap<SequenceHash, Weak<BlockHandle>>>>,
     event_manager: Arc<dyn EventManager>,
+    global_registry: GlobalRegistry,
+    unregister_tx: mpsc::UnboundedSender<SequenceHash>,
 }
 
 impl BlockRegistry {
-    pub fn new(event_manager: Arc<dyn EventManager>) -> Self {
+    pub fn new(
+        event_manager: Arc<dyn EventManager>,
+        global_registry: GlobalRegistry,
+        async_runtime: Handle,
+    ) -> Self {
+        let (unregister_tx, mut unregister_rx) = mpsc::unbounded_channel();
+
+        let blocks = Arc::new(Mutex::new(HashMap::new()));
+
+        let blocks_clone = blocks.clone();
+        let global_registry_clone = global_registry.clone();
+        async_runtime.spawn(async move {
+            let blocks = blocks_clone;
+            let global_registry = global_registry_clone;
+            while let Some(sequence_hash) = unregister_rx.recv().await {
+                {
+                    let mut blocks = blocks.lock().unwrap();
```

Comment:
do we need to acquire both locks together?

is this consistent with the other place we acquire both locks?